### PR TITLE
Add a section about how to add an argument to a method without breaking BC

### PR DESCRIPTION
**Pull request by fabpot**
Our section about BC breaks is too strict in the sense that we do have solutions for some of the things mentioned as not possible.
This PR documents one of them.

### CODE REVIEW

File: contributing/code/bc.rst [hunk 1]
```diff
@@ -334,84 +334,145 @@ Changing Traits
...
-Remove type hint of an argument                     No
-Change argument type                                No
-Change return type                                  No
+Add public method                                                                Yes
+Remove public method                                                             No
+Change name                                                                      No
+Reduce visibility                                                                No
+Make final                                                                       No              :ref:`[6] <note-6>`
+Move to used trait                                                               Yes
+:ref:`Add argument with a default value <add-argument-public-method>`            No
```
Comment (stof) [inline contributing/code/bc.rst]: changing `without` to `with` looks wrong to me

File: contributing/code/bc.rst [hunk 2]
```diff
@@ -334,84 +334,145 @@ Changing Traits
...
+Turn non static into static                                                      No
+Turn static into non static                                                      No
+===============================================================================  ==============  ===============
+
+Making Code Changes in a Backward Compatible Way
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As you read above, many changes are not allowed because they would represent a
+backward compability break. However, we want to be able to improve the code and
+its features over time and that can nbe done thanks to some strategies that
```
Comment (stof) [inline contributing/code/bc.rst]:
```suggestion
its features over time and that can be done thanks to some strategies that
```

File: contributing/code/bc.rst [hunk 3]
```diff
@@ -334,84 +334,149 @@ Changing Traits
...
+
+        if (\func_num_args() < 2) {
+            trigger_deprecation('symfony/COMPONENT', 'X.Y', 'The "%s()" method will have a new "bool $stripWithespace" argument in version Z.0, not defining it is deprecated.', __METHOD__);
+
+            $stripWithespace = false;
+        } else {
+            $stripWithespace = func_get_arg(1);
+        }
+
+5/ In the next major version (``X.0``), uncomment the argument, remove the
```
Comment (stof) [inline contributing/code/bc.rst]: shouldn't this be `#.` ?

File: contributing/code/bc.rst [hunk 4]
```diff
@@ -334,84 +334,149 @@ Changing Traits
...
+PHPDoc if there is no need for a description, and remove the ``func_get_arg``
+code and the warning if any.
 
 Notes
```
Comment (stof) [inline contributing/code/bc.rst:416]: shouldn't those notes stay below the table ?

File: contributing/code/bc.rst (hunk 1 above)
Comment (fabpot) [reply contributing/code/bc.rst]: oops

File: contributing/code/bc.rst (hunk 2 above)
Comment (fabpot) [reply contributing/code/bc.rst]: fixed

File: contributing/code/bc.rst (hunk 3 above)
Comment (fabpot) [reply contributing/code/bc.rst]: Yep

File: contributing/code/bc.rst (hunk 4 above)
Comment (fabpot) [reply contributing/code/bc.rst:416]: Sure

File: contributing/code/bc.rst [hunk 5]
```diff
@@ -473,4 +473,69 @@ a return type is only possible with a child type.
...
+        public function say(string $text, /* bool $stripWithespace = true */): void
+        {
+        }
+
+        // or required
+        public function say(string $text, /* bool $stripWithespace */): void
+        {
+        }
+
+#. Document the new argument in a PHPDoc::
```
Comment (GromNaN) [inline contributing/code/bc.rst:508]: We can mention this must be documented in all child classes.

File: contributing/code/bc.rst [hunk 6]
```diff
@@ -334,84 +334,84 @@ Changing Traits
...
-Change return type                                  No
+Add protected method                                                             Yes
+Remove protected method                                                          No
+Change name                                                                      No
+Reduce visibility                                                                No
+Make final                                                                       No              :ref:`[6] <note-6>`
+Make public                                                                      No              :ref:`[8] <note-8>`
+Move to used trait                                                               Yes
+Add argument without a default value                                             No
+Add argument with a default value                                                No
```
Comment (stof) [inline contributing/code/bc.rst]: those should probably be links too.

File: contributing/code/bc.rst [hunk 7]
```diff
@@ -334,84 +334,84 @@ Changing Traits
...
-Change argument type                                No
-Change return type                                  No
+Add public method                                                                Yes
+Remove public method                                                             No
+Change name                                                                      No
+Reduce visibility                                                                No
+Make final                                                                       No              :ref:`[6] <note-6>`
+Move to used trait                                                               Yes
+:ref:`Add argument without a default value <add-argument-public-method>`         No
+:ref:`Add argument with a default value <add-argument-public-method>`            No
```
Comment (stof) [inline contributing/code/bc.rst:369]: Should the text be a link or should we use a note, like for other cases where we have an alternative solution when sayign `No` ?.

Comment (stof) [reply contributing/code/bc.rst:369]: And we should also have similar links in the tables about working with classes (this is the table about traits)

**Review by stof — COMMENTED** [review]
We should probably also document the strategy to add new methods in an interface

File: contributing/code/bc.rst (hunk 6 above)
Comment (fabpot) [reply contributing/code/bc.rst]: added some more links

File: contributing/code/bc.rst (hunk 7 above)
Comment (fabpot) [reply contributing/code/bc.rst:369]: Done for the second one.
Not sure about the first comment.

**Comment by fabpot** [comment]
> We should probably also document the strategy to add new methods in an interface

Probably the next to document indeed.

File: contributing/code/bc.rst [hunk 8]
```diff
@@ -473,4 +473,69 @@ a return type is only possible with a child type.
...
+
+        // the new argument can be optional
+        public function say(string $text, /* bool $stripWithespace = true */): void
+        {
+        }
+
+        // or required
+        public function say(string $text, /* bool $stripWithespace */): void
+        {
+        }
```
Comment (OskarStark) [inline contributing/code/bc.rst]: Indention should be 4 spaces, instead of 8, no?

Comment (fabpot) [reply contributing/code/bc.rst]: done

**Comment by javiereguiluz** [comment]
Thank you Fabien.

**Pull request merged**